Installation issue: adios2@gcc%13

### Steps to reproduce the issue

```console
$ spack spec -I visit%gcc@13.2.0+osmesa ^openmpi@4.1.6
Input spec
--------------------------------
 -   visit%gcc@13.2.0+osmesa
 -       ^openmpi@4.1.6

Concretized
--------------------------------
 -   visit@3.3.3%gcc@13.2.0+adios2+conduit~gui+hdf5~ipo+mfem+mpi+netcdf+osmesa+plugins+python+silo~vtkm build_system=cmake build_type=Release generator=ninja patches=8fb2ef2,bd43c00 arch=linux-rhel8-haswell
 -       ^adios2@2.7.1%gcc@13.2.0+blosc+bzip2~cuda~dataspaces+fortran+hdf5~ipo+mpi~pic+png+python~rocm+ssc+sst+sz+zfp build_system=cmake build_type=Release generator=make patches=8221073,88b2cd1 arch=linux-rhel8-haswell
...
```


### Error message

```console
==> Installing adios2-2.7.1-2q274aiw6jm6vpzmoplyrsrkj6st7mt7 [109/121]
==> No binary for adios2-2.7.1-2q274aiw6jm6vpzmoplyrsrkj6st7mt7 found: installing from source
==> Using cached archive: /vast/home/ptb/spack/var/spack/cache/_source-cache/archive/c8/c8e237fd51f49d8a62a0660db12b72ea5067512aa7970f3fcf80b70e3f87ca3e.tar.gz
==> Using cached archive: /vast/home/ptb/spack/var/spack/cache/_source-cache/archive/82/8221073d1b2f8944395a88a5d60a15c7370646b62f5fc6309867bbb6a8c2096c
==> Applied patch https://github.com/ornladios/adios2/pull/2710.patch?full_index=1
==> Applied patch /vast/home/ptb/spack/var/spack/repos/builtin/packages/adios2/2.7.1-fix-broken-endian-reverse-compile.patch
==> adios2: Executing phase: 'cmake'
==> adios2: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j10'

107 errors found in build log:
     1500    cd /tmp/ptb/spack-stage/spack-stage-adios2-2.7.1-2q274aiw6jm6vpzmoplyrsrkj6st7mt7/spack-build-2q274ai/bindings/Fortran && /vast/home/ptb/spack/opt/spack/linux-rhel8-h
             aswell/gcc-13.2.0/cmake-3.27.9-wsrith7lvezc6d5c3sd57u7ig5rngjke/bin/cmake -E cmake_link_script CMakeFiles/adios2_fortran_mpi.dir/link.txt --verbose=1
     1501    /vast/home/ptb/spack/lib/spack/env/gcc/g++ -fPIC -O3 -DNDEBUG -Wl,-rpath -Wl,/projects/opt/rhel8/x86_64/openmpi/4.1.6-gcc_13.2.0/lib -L/vast/projects/opt/rhel8/x86_64
             /openmpi/4.1.6-gcc_13.2.0/lib -pthread -shared -Wl,-soname,libadios2_fortran_mpi.so.2 -o ../../lib64/libadios2_fortran_mpi.so.2.7.1 CMakeFiles/adios2_fortran_mpi.dir/
             modules/adios2_adios_init_mpi_smod.F90.o CMakeFiles/adios2_fortran_mpi.dir/modules/adios2_io_open_mpi_smod.F90.o CMakeFiles/adios2_fortran_mpi.dir/f2c/adios2_f2c_adio
             s_mpi.cpp.o CMakeFiles/adios2_fortran_mpi.dir/f2c/adios2_f2c_io_mpi.cpp.o  -Wl,-rpath,/tmp/ptb/spack-stage/spack-stage-adios2-2.7.1-2q274aiw6jm6vpzmoplyrsrkj6st7mt7/s
             pack-build-2q274ai/lib64:/vast/home/ptb/spack/opt/spack/linux-rhel8-haswell/gcc-13.2.0/zfp-0.5.5-knadzofra42yv4s3qmwcff3vbtckbdom/lib64:/projects/opt/rhel8/x86_64/ope
             nmpi/4.1.6-gcc_13.2.0/lib: ../../lib64/libadios2_fortran.so.2.7.1 ../../lib64/libadios2_c_mpi.so.2.7.1 ../../lib64/libadios2_core_mpi.so.2.7.1 ../../lib64/libadios2_c
             .so.2.7.1 ../../lib64/libadios2_core.so.2.7.1 -lpthread /projects/opt/rhel8/x86_64/openmpi/4.1.6-gcc_13.2.0/lib/libmpi_usempif08.so /projects/opt/rhel8/x86_64/openmpi
             /4.1.6-gcc_13.2.0/lib/libmpi_usempi_ignore_tkr.so /projects/opt/rhel8/x86_64/openmpi/4.1.6-gcc_13.2.0/lib/libmpi_mpifh.so /projects/opt/rhel8/x86_64/openmpi/4.1.6-gcc
             _13.2.0/lib/libmpi.so -lgfortran -lquadmath -Wl,-rpath-link,/tmp/ptb/spack-stage/spack-stage-adios2-2.7.1-2q274aiw6jm6vpzmoplyrsrkj6st7mt7/spack-build-2q274ai/lib64:/
             vast/home/ptb/spack/opt/spack/linux-rhel8-haswell/gcc-13.2.0/zfp-0.5.5-knadzofra42yv4s3qmwcff3vbtckbdom/lib64
     1502    In file included from /tmp/ptb/spack-stage/spack-stage-adios2-2.7.1-2q274aiw6jm6vpzmoplyrsrkj6st7mt7/spack-src/thirdparty/pybind11/pybind11/include/pybind11/pybind11.
             h:45,
     1503                     from /tmp/ptb/spack-stage/spack-stage-adios2-2.7.1-2q274aiw6jm6vpzmoplyrsrkj6st7mt7/spack-src/thirdparty/pybind11/pybind11/include/pybind11/numpy.h:1
             2,
     1504                     from /tmp/ptb/spack-stage/spack-stage-adios2-2.7.1-2q274aiw6jm6vpzmoplyrsrkj6st7mt7/spack-src/bindings/Python/py11IO.h:14,
     1505                     from /tmp/ptb/spack-stage/spack-stage-adios2-2.7.1-2q274aiw6jm6vpzmoplyrsrkj6st7mt7/spack-src/bindings/Python/py11IO.cpp:11:
  >> 1506    /tmp/ptb/spack-stage/spack-stage-adios2-2.7.1-2q274aiw6jm6vpzmoplyrsrkj6st7mt7/spack-src/thirdparty/pybind11/pybind11/include/pybind11/attr.h:199:10: error: 'uint16_t
             ' in namespace 'std' does not name a type; did you mean 'wint_t'?
```

### Information on your system

* **Spack:** 0.22.0.dev0 (2b3e7fd10a75adb73eeb8465af265da9ffddcfc7)
* **Python:** 3.6.8
* **Platform:** linux-rhel8-haswell
* **Concretizer:** clingo

### Additional information

maintainers: @ax3l  @vicentebolea  @williamfgc
Perhaps a change in transitive includes occurred in gcc13

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

As a note, this is unrelated to our just released adios 2.10.0 since it is using adios2 2.7.1 as shown in the log:
```
  >> 1506    /tmp/ptb/spack-stage/spack-stage-adios2-2.7.1-2q274aiw6jm6vpzmoplyrsrkj6st7mt7/spack-src/thirdparty/pybind11/pybind11/include/pybind11/attr.h:199:10: error: 'uint16_t
  
  ```

--- a/var/spack/repos/builtin/packages/adios2/2.7-fix-missing-cstdint-include.patch
+++ b/var/spack/repos/builtin/packages/adios2/2.7-fix-missing-cstdint-include.patch
@@ -1,0 +1,24 @@
+From abbe0f81d78403a62444b1a76222b9a226fd17c6 Mon Sep 17 00:00:00 2001
+From: Vicente Adolfo Bolea Sanchez <vicente.bolea@kitware.com>
+Date: Mon, 22 Apr 2024 20:52:49 -0700
+Subject: [PATCH] fix py11bind stdint dep
+
+---
+ thirdparty/pybind11/pybind11/include/pybind11/pybind11.h | 1 +
+ 1 file changed, 1 insertion(+)
+
+diff --git a/thirdparty/pybind11/pybind11/include/pybind11/pybind11.h b/thirdparty/pybind11/pybind11/include/pybind11/pybind11.h
+index fa98f7c18..6992c3d55 100644
+--- a/thirdparty/pybind11/pybind11/include/pybind11/pybind11.h
++++ b/thirdparty/pybind11/pybind11/include/pybind11/pybind11.h
+@@ -47,6 +47,7 @@
+ #include "detail/class.h"
+ #include "detail/init.h"
+ 
++#include <cstdint>
+ #include <memory>
+ #include <vector>
+ #include <string>
+-- 
+2.35.3
+

--- a/var/spack/repos/builtin/packages/adios2/package.py
+++ b/var/spack/repos/builtin/packages/adios2/package.py
@@ -209,6 +209,9 @@ class Adios2(CMakePackage, CudaPackage, ROCmPackage):
     # See https://github.com/ornladios/ADIOS2/pull/2714
     patch("2.6-fix-gcc10-symbols.patch", when="@2.6.0")
 
+    # add missing include <cstdint>
+    patch("2.7-fix-missing-cstdint-include.patch", when="@2.7")
+
     # Add missing include <memory>
     # https://github.com/ornladios/adios2/pull/2710
     patch(

